#Importing libraries

In [2]:
#Import libraries
import pandas as pd
import numpy as np
import os

In [9]:
#Importing orders.csv
path =r'C:\Users\Anwender\Instacart Basket Analysis'
df_ords= pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [4]:
#Importing products.csv
path =r'C:\Users\Anwender\Instacart Basket Analysis'
df_prods= pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [6]:
#Create a test dataframe
df_test=pd.DataFrame()

In [7]:
#Create a mix type column
df_test['mix']=['a','b',1, True]

In [8]:
#check what is inside df_test
df_test.head()

mix
0     a
1     b
2     1
3  True

In [11]:
#checking for mixed data types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [12]:
df_test['mix'] = df_test['mix'].astype('str')

# Finding missing values

In [13]:

df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [14]:
#creating a dataframe only with null values from the product_name column
df_nan=df_prods[df_prods['product_name'].isnull()==True]

In [15]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [16]:
#comparing number of rows in current dataframe versus subset
df_prods.shape

(49693, 5)

In [17]:
#creating new dataframe without the rows with missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [19]:
#checking new dataframe
df_prods_clean.shape

(49677, 5)

In [22]:
#dropping NaNs from the original file
df_prods.dropna(inplace = True)


# Finding Duplicates

In [24]:
#finding duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [25]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

# Addressing duplicates

In [26]:
#check current numbers of rows in df_prods_clean
df_prods_clean.shape

(49677, 5)

In [27]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [28]:
df_prods_clean_no_dups.shape

(49672, 5)

# Task 2 - Run the df.describe() function on your df_ords dataframe.

In [31]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

## 1 - the 'Unnamed' column probably should have a name as I have no idea what am I looking at
## 2 - all these long numbers like 3.3456e+06 do not make much sense to me. Especially in orders_day_of_week and order_hour_of_day, they should be integers, while in here we have 3.42 for example.


# Task 3 - Check for mixed-type data in your df_ords dataframe.



In [33]:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col, ' mixed')
  else: print (col, ' consistent')

Unnamed: 0  consistent
order_id  consistent
user_id  consistent
order_number  consistent
orders_day_of_week  consistent
order_hour_of_day  consistent
days_since_prior_order  consistent


## no mixed type data

# Task 5- Run a check for missing values in your df_ords dataframe. 

In [37]:
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

## we have missing values only in the days_since_prior_order column. I think this is ok, since all the new users that buy for the first time will have no prior orders.

# Task 6- addressing missing values

## I will not be filling these missing values, as they do represent that those users are first-time users.

# Task 7 - Run a check for duplicate values in your df_ords data

In [39]:
# Run a check for duplicate values in your df_ords data.

df_ords_dups = df_ords[df_ords.duplicated()]

In [40]:
# Visualize df_ords_dups

df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

## no duplicates

# Task 9 - Export everything

In [43]:
# Export df_prods_clean_no_dups

df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

In [44]:
# Export final df_ords 

df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))